mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java

```diff
@@ -0,0 +1,117 @@
+        this.ctx = ctx;
+        this.settings = settings;
+    }
+
+    protected void updateHttpClientAuth(MyHttpClient httpClient) {
+        if ((getUsername() != null && !getUsername().isEmpty()) ||
+                (getPassword() != null && !getPassword().isEmpty())) {
+            httpClient.setBasicAuth(getUsername(), getPassword());
+        }
+    }
+    public MyAsyncHttpClient getAsyncHttpClient() {
+        if (asyncHttpClient == null) {
+            asyncHttpClient = new MyAsyncHttpClient(ctx, ignoreSslHostname(), ignoreCertTrust());
+            asyncHttpClient.setTimeout(30000);
+            asyncHttpClient.setBaseUrl(getOpenHABUrl());
+
+            updateHttpClientAuth(asyncHttpClient);
+        }
+        return asyncHttpClient;
+    }
+
+    private Boolean ignoreCertTrust() {
+        return settings.getBoolean(Constants.PREFERENCE_SSLCERT, false);
+    }
+
+    private Boolean ignoreSslHostname() {
+        return settings.getBoolean(Constants.PREFERENCE_SSLHOST, false);
+    };
+
+    public MySyncHttpClient getSyncHttpClient() {
+        if (syncHttpClient == null) {
+            syncHttpClient = new MySyncHttpClient(ctx, ignoreSslHostname(), ignoreCertTrust());
+            syncHttpClient.setBaseUrl(getOpenHABUrl());
+
+            updateHttpClientAuth(syncHttpClient);
+        }
+        return syncHttpClient;
+    }
+
+    @Override
+    public void setConnectionType(int connectionType) {
+        this.connectionType = connectionType;
+    }
+
+    @Override
+    public int getConnectionType() {
+        return this.connectionType;
+    }
+
+    @Override
+    public boolean isReachable() {
+        Log.d(TAG, "Checking reachability of " + getOpenHABUrl());
+        try {
+            AsyncTask task = new AsyncTask<String, Void, Boolean>() {
+                @Override
+                protected Boolean doInBackground(String... strings) {
+                    try {
+                        URL url = new URL(strings[0]);
+                        int checkPort = url.getPort();
+                        if (url.getProtocol().equals("http") && checkPort == -1)
+                            checkPort = 80;
+                        if (url.getProtocol().equals("https") && checkPort == -1)
+                            checkPort = 443;
+                        Socket s = new Socket();
+                        s.connect(new InetSocketAddress(url.getHost(), checkPort), 1000);
+                        Log.d(TAG, "Socket connected");
+                        s.close();
+                        return true;
+                    } catch (Exception e) {
+                        Log.e(TAG, e.getMessage());
+                        return false;
+                    }
+                }
+            }.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, getOpenHABUrl());
+
+            // task can be null, e.g. unit tests, assume the target is reachable in this case
+            return task == null || (boolean) task.get();
```

Comment:
I "just" copied this part from the original code without thinking about that. However, now when you say that, it seems it's pointless doing it in a separate thread. I'll remove that :)

Reply:
I _think_ I know why the original author chose to do this: It'll avoid getting NetworkOnMainThreadException as the network I/O technically is done in the AsyncTask thread. However, this seems like a hack because it certainly doesn't follow the spirit of 'don't do IO on main thread', as the main thread is blocked until the IO is finished.

This isReachable() method should probably better be asynchronous or calls to it moved to a background thread. Not sure how to achieve this with the current design though :-(

Reply:
So, this is a problem for each architecture, where we've this selection method of servers (the isReachable method is only used to check, if the local connection is reachble or not), as this would always be a synchronous step, isn't it? :/

Reply:
Not if a background thread is used to
- monitor connectivity
- monitor reachability state
- select the active connection, which is then returned by getConnection

Reply:
The app, how it's done currently, requires to have a connection selected when it is started. It would probably be possible to implement such a background thread for the time the app is already running (to monitor changes the connectivity when the app is used). However, the first selection needs to be synchronous, wouldn't it?

Reply:
Strictly speaking yes, but making it 'semi synchronous' would probably be fine as well: Fire up the thread when the app is started and show a spinning wheel during the (likely very short) time until the initial decision what connection to use is made.